### ISSUE TITLE
Add text drawing

### ISSUE THREAD

**Issue by faiface** [faiface/pixel#12]
Currently, Pixel doesn't support drawing text. This needs to be changed. Implementing text drawing in Go is actually way simpler than I thought it would be.

First of all, there's an awesome package https://godoc.org/golang.org/x/image/font, which implements standard font face interface, which is well designed and very useful. Subpackages implement several font types and https://godoc.org/github.com/golang/freetype/truetype implements TrueType font loading with support of the standard font face interface. So we've got all we need.

Now, the design of text drawing in Pixel.

Text drawing should be done in a not-yet-existing `"github.com/faiface/pixel/text"` package. This package should export (at least) one type: `Text`. Creating one should be done by `text.New(face)`, where `face` is a `font.Face` value. The `*Text` value returned from `New` will use that specific font face.

`Text` type should implement `io.Writer` interface, so drawing text to it is as simple as `fmt.Sprintf(txt, "hello world")`. Printing to `Text` draws text "to the `Text` object". To draw the printed text to another target (such as a window, canvas, batch, ...), simply use `txt.Draw(target)`. Clearing needs to be supported too (`txt.Clear()`).

Drawing text needs to be fast. There should be no problem drawing a full screen of random text using a tiny font face at 60 FPS without a performance hit. This is something that many other libraries fail to accomplish, but it's actually not very hard. It's easily achievable by caching glyphs.

So, to recap the interface.

- `"github.com/faiface/pixel/text"`
  - `New(font.Face) *Text`
  - `type Text`
    - `Clear()`
    - `Draw(pixel.Target)`
    - `Write(b []byte) (n int, err error)`

This is the basis for the interface. More methods will probably be added, such as `WriteString`, or `SetDot` (the position of the text line).

**Comment by Chillance** [comment]
It might also be handy with a GetSize, that returns the size the text takes (width, height) in pixels. Handy if you want to figure out and layout things.

**Comment by faiface** [comment]
@Chillance Good point. I'd suggest a method called, for example `Bounds(string) pixel.Rect`, which would return a bounding box of a string if it was to be written right now. Another option would be a method `Bounds() pixel.Rect`, which would return the bounding box of the text that's currently written in the `Text` object.

The height of the returned rectangle would be the line height, not the actual height of the text (which would differ for lower case and upper case letters), in order to avoid text misaligning.

**Comment by Chillance** [comment]
Yeah, sounds good to me.

**Comment by faiface** [comment]
@Chillance Which one do you think would be more useful? The one that calculates the bounding box of a string that's "to be written", or the one which calculates the bounding box of the "already written text". Note, that "already written text" is just written to the `Text` object so it still can be drawn to an arbitrary location on the screen.

**Comment by Chillance** [comment]
Well, I suppose the first one would help more in the layouting scenario, as then you can check the bounds of the text before actually writing it, and thus you can position it properly when you do.

The other one might still be handy too. So, why not both? :)

**Comment by faiface** [comment]
That's right, I also see both of them being useful, so now it comes to a naming problem. I'll figure that out :)

**Comment by faiface** [comment]
@Chillance So, to explain what the difference would be and how this would work (this assumes you know how to use the rest of the library). Here's a basic usecase:

```go
txt := text.New(face)
txt.SetDot(pixel.V(5, 3))
fmt.Sprint(txt, "hello ", 12, "!")
txt.Draw(win)
txt.SetMatrix(pixel.IM.Moved(pixel.V(100, 0))
txt.Draw(win)
```

In this example, we print a text to the `Text` object. The text is then saved inside the `Text` object. Then, we draw the text two times to the window, first time to the original "dot position". The second time shifted to the right by 100 pixels.

Does it make everything clearer?

**Comment by Chillance** [comment]
Aha, reusing the text. No animation here. :)

In my case I would just like to know how much space a certain text takes so I can position it properly. Handy when doing my own little text widgets for example.

**Comment by faiface** [comment]
You can achieve animation by drawing the same text at a different location each frame, if that's what you meant ;)

**Comment by faiface** [comment]
And yeah, so the first alternative would enable you to examine the size of a text before printing it a `txt` variable and adjust the `txt`'s dot accordingly. With the second alternative, you'd first print the text to the `txt`, then check it's size and finally position it correctly on the screen.

**Comment by Chillance** [comment]
Animation was just a quick remark that interpolating between two positions build into the library would be cool, but yes, I suppose it can be done by user too...

**Comment by rucuriousyet** [comment]
I might be able to help with this. I have some basic code for getting text running thx to some friends. It needs some work but should be enough to help get stuff started.

**Comment by faiface** [comment]
@rucuriousyet I've already started working on this, so no help is currently required. Testing and finding bugs will be useful once I push my changes.

**Comment by faiface** [comment]
Hey everyone!

I created a `text` branch and pushed the first preliminary feature-incomplete `Text` type implementation for drawing text. It works fairly well from what I've tried.

Test it out, test it to the limits, find bugs, report them, thank you 🍦 .

So, here's a quick tut on how to use it, since there are no doc comments yet.

The new package is `"github.com/faiface/pixel/text"` and the most important type inside it is `Text`.

To create `Text` object, use `text.New` constructor, which takes a few arguments. The first one is a font face that you want to use with that text object. The simplest one available is [here](https://godoc.org/golang.org/x/image/font/basicfont#pkg-variables). You can also load font faces from TTF files, here's a quick function which does that ([font](https://godoc.org/golang.org/x/image/font) and [truetype](https://godoc.org/github.com/golang/freetype/truetype) packages required).

```go
func loadTTF(path string, options *truetype.Options) (font.Face, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	b, err := ioutil.ReadAll(file)
	if err != nil {
		return nil, err
	}

	ttf, err := truetype.Parse(b)
	if err != nil {
		return nil, err
	}

	return truetype.NewFace(ttf, options), nil
}
```

Following the face argument, `text.New` constructor takes variable number of rune sets, which are just slices of runes. The returned `Text` will only be able to draw runes contained in the provided sets. The `text` package contains a predefined `ASCII` rune set, which contains all printable ASCII runes. There's also a convenient `RangeTable` function, which converts `unicode.RangeTable` to a rune set. So, here's an example:

```go
// ...
// load a face here
// ...
txt := text.New(face, text.ASCII)
```

And here's an example that can also draw Japanese (if you've got some font for that):

```go
txt := text.New(face, text.ASCII, text.RangeTable(unicode.Katakana))
```

After creation, the `Text` object can be used. Since it implements `io.Writer` interface, you can write to it using standard functions from the `"fmt"` package, such as `fmt.Fprintf`. If you want to remove all of the text written to the `Text` object, use `txt.Clear()`.

Note, that a `Text` object is a kind of a buffer. Writing to it doesn't directly draw the text anywhere. To draw the buffered text, use `txt.Draw(target)` (for example, `txt.Draw(win)`).

Currently, the only way to control the position of the drawn text is to set `txt.Dot` position directly, or use `target.SetMatrix`.

What's missing is an easy what to move the text without moving the dot or setting the matrix of a target. Also, it might be cool to be able to control the way the dot advances, so we'd be able to draw a text sitting on a curved line, just thoughts.

So, test and report bugs and provide suggestions on improvements!

**Comment by rucuriousyet** [comment]
Cool. I will be starting a gui project using Pixel sometime in the next few weeks. I figured it would be best to start by writing an application rather than a library, So I am going to start by writing something like Sublime or Atom. After some progress is made I will probably release the components as a library. Let me know if you have any interest in helping out.

**Comment by rucuriousyet** [comment]
Im starting out with a more JavaFX style of GUI programming but I hope to eventually have support for something like HTML and CSS (or maybe a custom language).

**Comment by faiface** [comment]
@rucuriousyet Cool, but this belongs to #17, this issue is just about text drawing. You, we want everything organized to some degree at least :)

**Comment by faiface** [comment]
I've added `Text.SetMatrix` and `Text.SetColorMask`. They work similarly as with `Sprite`. You can write some text, then call `txt.SetMatrix` and it will affect all written text.

**Comment by faiface** [comment]
Just added full `text.Atlas` and `text.Text` documentation.

**Comment by faiface** [comment]
Text drawing is implemented and documented. Any issues regarding it should be reported in new issues.
